File: src/areaint.f
C Output from Public domain Ratfor, version 1.0
      subroutine areaint(u,v,ix,x,y,n,par,period,cifval)
      implicit double precision(a-h,o-z)
      dimension par(3), x(1), y(1), period(2)
      logical per
      logical covered
      logical neigh(n)
      eps = 2.22d-16
      zero = 0.d0
      ngrid = 64
      per = period(1) .gt. zero
      beta = par(1)
      eta = par(2)
      r = par(3)
      if(n.eq.0)then
      cifval = beta
      return
      endif
      r2 = r * r
      dx = (2 * r)/dble(ngrid)
      dy = dx
      range2 = 4 * r2
      if(.not. per)then
      ixm1 = ix - 1
      ixp1 = max(1, ix + 1)
      if(ixm1 .gt. 0)then
      do23006 j = 1,ixm1 
      a = range2 - (u - x(j))**2
      neigh(j) = .false.
      if(a .gt. 0)then
      a = a - (v - y(j))**2
      if(a .gt. 0)then
      neigh(j) = .true.
      endif
      endif
23006 continue
23007 continue
      endif
      if(ixp1 .le. n)then
      do23014 j = ixp1,n 
      a = range2 - (u - x(j))**2
      neigh(j) = .false.
      if(a .gt. 0)then
      a = a - (v - y(j))**2
      if(a .gt. 0)then
      neigh(j) = .true.
      endif
      endif
23014 continue
23015 continue
      endif
      kount = 0
      kdisc = 0
      xgrid0 = u - r - dx/2
      do23020 kx=1,ngrid 
      xgrid = xgrid0 + dble(kx) * dx
      my = int(dsqrt(r2 - (u - xgrid)**2)/dy)
      if(my .ge. 0)then
      do23024 ky=-my,my 
      ygrid = v + dble(ky) * dy
      kdisc = kdisc + 1
      covered = .false.
      if(ixm1 .gt. 0)then
      do23028 j = 1,ixm1 
      if(neigh(j))then
      a = r2 - (xgrid - x(j))**2
      if(a .gt. 0)then
      a = a - (ygrid - y(j))**2
      if(a .gt. 0)then
      covered = .true.
      goto 42
      endif
      endif
      endif
23028 continue
23029 continue
      endif
      if(ixp1 .le. n)then
      do23038 j = ixp1,n 
      if(neigh(j))then
      a = r2 - (xgrid - x(j))**2
      if(a .gt. 0)then
      a = a - (ygrid - y(j))**2
      if(a .gt. 0)then
      covered = .true.
      goto 42
      endif
      endif
      endif
23038 continue
23039 continue
      endif
      if(.not. covered)then
      kount = kount + 1
      endif
42    continue
23024 continue
23025 continue
      endif
23020 continue
23021 continue
      else
      ixm1 = ix - 1
      ixp1 = max(1, ix + 1)
      if(ixm1 .gt. 0)then
      do23050 j = 1,ixm1 
      call dist2(u,v,x(j),y(j),period,d2)
      neigh(j) = (d2 .lt. range2)
23050 continue
23051 continue
      endif
      if(ixp1 .le. n)then
      do23054 j = ixp1,n 
      call dist2(u,v,x(j),y(j),period,d2)
      neigh(j) = (d2 .lt. range2)
23054 continue
23055 continue
      endif
      kount = 0
      kdisc = 0
      xgrid0 = u - r - dx
      ygrid0 = v - r - dy
      do23056 kx=1,ngrid 
      xgrid = xgrid0 + dble(kx) * dx
      do23058 ky=1,ngrid 
      ygrid = ygrid0 + dble(ky) * dy
      call dist2(u,v,xgrid,ygrid,period,d2)
      if(d2 .lt. r2)then
      kdisc = kdisc + 1
      covered = .false.
      if(ixm1 .gt. 0)then
      do23064 j = 1,ixm1 
      if(neigh(j))then
      call dist2(xgrid,ygrid,x(j),y(j),period,d2)
      if(d2 .lt. r2)then
      covered = .true.
      goto 4242
      endif
      endif
23064 continue
23065 continue
      endif
      if(ixp1 .le. n)then
      do23072 j = ixp1,n 
      if(neigh(j))then
      call dist2(xgrid,ygrid,x(j),y(j),period,d2)
      if(d2 .lt. r2)then
      covered = .true.
      goto 4242
      endif
      endif
23072 continue
23073 continue
      endif
      if(.not. covered)then
      kount = kount + 1
      endif
4242  continue
      endif
23058 continue
23059 continue
23056 continue
23057 continue
      endif
      if(eta .gt. eps)then
      covfrac = dble(kdisc - kount) / dble(kdisc)
      cifval = beta * exp(log(eta) * covfrac)
      else
      if(kount .eq. kdisc)then
      cifval = beta
      else
      cifval = zero
      endif
      endif
      return
      end

File: src/areaint.c
#include <R.h>
#include <math.h>
#include <stdlib.h>
#include "methas.h"
#include "dist2.h"

#define NGRID 16

/*
  Conditional intensity function for an area-interaction process:

  cif = eta^(1-B) where B = (uncovered area)/(pi r^2)

*/

/* Format for storage of parameters and precomputed/auxiliary data */

typedef struct AreaInt {
  /* model parameters */
  double eta;
  double r;
  /* transformations of the parameters */
  double r2;
  double range2;
  double logeta;
  double hard;
  /* periodic distance */
  double *period;
  int per;
  /* grid counting */
  double dx;
  double xgrid0;
  int *my;
  int kdisc;
  /* auxiliary indicators */
  char *neigh;
} AreaInt;

/* initialiser function */

Cdata *areaintInit(state, model, algo)
     State state;
     Model model;
     Algor algo;
{
  double r, dx, dy, x0;
  int i, my, kdisc;
  AreaInt *areaint;

  /* create storage */
  areaint = (AreaInt *) R_alloc(1, sizeof(AreaInt));
  /* Interpret model parameters*/
  areaint->eta    = model.ipar[0];
  areaint->r      = r = model.ipar[1]; 
  areaint->r2     = r * r;
  areaint->range2 = 4 * r * r;    /* square of interaction distance */
  /* is the model numerically equivalent to hard core ? */
  areaint->hard   = (areaint->eta < DOUBLE_EPS);
  areaint->logeta = (areaint->hard) ? 0 : log(areaint->eta);
  /* periodic boundary conditions? */
  areaint->period = model.period;
  areaint->per    = (model.period[0] > 0.0);
  /* grid counting */
  dx = dy = areaint->dx = (2 * r)/NGRID;
  areaint->xgrid0 = -r + dx/2;
  areaint->my = (int *) R_alloc((long) NGRID, sizeof(int));
  kdisc = 0;
  for(i = 0; i < NGRID; i++) {
    x0 = areaint->xgrid0 + i * dx;
    my = floor(sqrt(r * r - x0 * x0)/dy);
    my = (my < 0) ? 0 : my;
    areaint->my[i] = my;
    kdisc += 2 * my + 1;
  }
  areaint->kdisc = kdisc;
  /* allocate space for neighbour indicators */
  areaint->neigh = (char*) R_alloc((long) state.npmax, sizeof(char));
  return((Cdata *) areaint);
}

/* conditional intensity evaluator */

double areaintCif(prop, state, cdata)
     Propo prop;
     State state;
     Cdata *cdata;
{
  int npts, kount, ix, ixp1, j;
  double *period, *x, *y;
  double u, v;
  double r2, d2, dx, dy, range2;
  double a, xgrid0, ygrid, ygrid0, xgrid, covfrac, cifval;
  int kdisc, kx, my, ky, covered;
  AreaInt *areaint;

  areaint = (AreaInt *) cdata;

  period = areaint->period;
  r2     = areaint->r2;

  u  = prop.u;
  v  = prop.v;
  ix = prop.ix;
  x  = state.x;
  y  = state.y;

  npts = state.npts;
  if(npts == 0) return ((double) 1.0);

  r2 = areaint->r2;
  dy = dx = areaint->dx;
  range2 = areaint->r2;    /* square of interaction distance */

  if(!areaint->per) {
    /*
      Euclidean distance
      First identify which data points are neighbours of (u,v)
    */
    ixp1 = ix + 1;
    /* If ix = NONE = -1, then ixp1 = 0 is correct */
    if(ix > 0) {
      for(j=0; j < ix; j++) {
	a = range2 - pow(u - x[j], 2);
	areaint->neigh[j] = FALSE;
	if(a > 0.) {
	  a -= pow(v - y[j], 2);
	  if(a > 0.)
	    areaint->neigh[j] = TRUE;
	}
      }
    }
    if(ixp1 < npts) {
      for(j=ixp1; j < npts; j++) {
	a = range2 - pow(u - x[j], 2);
	areaint->neigh[j] = FALSE;
	if(a > 0.) {
	  a -= pow(v - y[j], 2);
	  if(a > 0.)
	    areaint->neigh[j] = TRUE;
	}
      }
    }
    /* scan a grid of points centred at (u,v) */
    kount = 0;
    xgrid0 = u + areaint->xgrid0;
    for(kx=0; kx<NGRID; kx++) {
      xgrid = xgrid0 + kx * dx;
      my = areaint->my[kx];
      for(ky=(-my); ky<=my; ky++) {
	ygrid = v + ky * dy;
	/*
	  Grid point (xgrid,ygrid) is inside disc of
	  radius r centred at (u,v)

	  Loop through all data points to determine
	  whether the grid point is covered by another disc
	*/
	covered = FALSE;
	if(ix > 0) {
	  for(j=0; j < ix; j++) {
	    if(areaint->neigh[j]) {
	      a = r2 - pow(xgrid - x[j], 2);
	      if(a > 0) {
		a -= pow(ygrid - y[j], 2);
		if(a > 0) {
		  /* point j covers grid point */
		  covered = TRUE;
		}
	      }
	    }
	  }
	}
	if(!covered && ixp1 < npts) {
	  for(j=ixp1; j<npts; j++) {
	    if(areaint->neigh[j]) {
	      a = r2 - pow(xgrid - x[j], 2);
	      if(a > 0) {
		a -= pow(ygrid - y[j], 2);
		if(a > 0) {
		  /* point j covers grid point */
		  covered = TRUE;
		}
	      }
	    }
	  }
	}
	/* finished scanning all data points j  */
	if(!covered)
	  kount = kount + 1;
	/* finished consideration of grid point (xgrid, ygrid) */
      }
    }
    kdisc = areaint->kdisc;
  } else {
    /*
      periodic distance
      First identify which data points are neighbours of (u,v)
    */
    ixp1 = ix + 1;
    if(ix > 0) {
      for(j=0; j < ix; j++) {
	d2 = dist2(u,v,x[j],y[j],period);
	areaint->neigh[j] = (d2 < range2);
      }
    }
    if(ixp1 < npts) {
      for(j=ixp1; j<npts; j++) {
	d2 = dist2(u,v,x[j],y[j],period);
	areaint->neigh[j] = (d2 < range2);
      }
    }
    /* scan a grid of ngrid * ngrid points centred at (u,v) */
    kount = 0;
    kdisc = 0;
    xgrid0 = u + areaint->xgrid0;
    ygrid0 = u + areaint->xgrid0;
    for(kx=0; kx<NGRID; kx++) {
      xgrid = xgrid0 + kx * dx;
      for(ky=0; ky<NGRID; ky++) {
	ygrid = ygrid0 + ky * dy;
	/* Determine whether grid point is inside disc */
	d2 = dist2(u,v,xgrid,ygrid,period);
	if(d2 < r2) {
	  /* Grid point is inside disc of radius r centred at (u,v) */
	  kdisc = kdisc + 1;
	  /*
	    Loop through all data points to determine
	    whether the grid point is covered by another disc
	  */
	  covered = FALSE;
	  if(ix > 0) {
	    for(j=0; j< ix; j++) {
	      if(areaint->neigh[j]) {
		d2 = dist2(xgrid,ygrid,x[j],y[j],period);
		if(d2 < r2) {
		  /* point j covers grid point */
		  covered = TRUE;
		}
	      }
	    }
	  }
	  if(!covered && ixp1 < npts) {
	    for(j=ixp1; j<npts; j++) {
	      if(areaint->neigh[j]) {
		d2 = dist2(xgrid,ygrid,x[j],y[j],period);
		if(d2 < r2) {
		  /* point j covers grid point */
		  covered = TRUE;
		}
	      }
	    }
	  }
	  /* finished scanning all data points j */
	  if(!covered) kount = kount + 1;
	  /* finished considering grid point (xgrid,ygrid) */
	}
      }
    }
  }

  /*
    `kdisc' is the number of           grid points in the disc
    `kount' is the number of UNCOVERED grid points in the disc
  */

  if(areaint->hard) {
    if(kount == kdisc) cifval = 1.0;
    else cifval = 0.0;
  } else {
    /* usual calculation
       COVERED area fraction */
    covfrac = ((double) kdisc - (double) kount)/((double) kdisc);
    cifval = exp(areaint->logeta * covfrac);
  }

  return cifval;
}


Cifns AreaIntCifns = { &areaintInit, &areaintCif, (updafunptr) NULL, FALSE};
